### PR TITLE
Add HLD for FRR-SONiC Communication Channel Enhancements

### PR DESCRIPTION
**Pull request by cscarpitta**
The document that describes the HLD is available here:
https://github.com/cscarpitta/SONiC/blob/frr-sonic-communication-channel-enhancements/doc/sonic-fpm-module/frr_sonic_communication_channel.md

| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-buildimage | [Introduce a SONiC-specific communication channel between FRR and SONiC](https://github.com/sonic-net/sonic-buildimage/pull/18715) | GitHub issue/pull request detail |
| sonic-swss | [[fpmsyncd]: Add support for SRv6](https://github.com/sonic-net/sonic-swss/pull/3123) | GitHub issue/pull request detail |

### CODE REVIEW

**Review by zice312963205 — APPROVED** [review]
Looks good

**Comment by prsunny** [comment]
@cscarpitta , can you please check the following?
1. Add a section in HLD for Warmboot implications.
2. Any modifications to existing APP_DB schema with this proposal?
3. Any sequencing issues causing dependency since now some are coming via netlink and some via protobuf?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+The lack of flexibility of Netlink in supporting various use cases in SONiC was previously discussed in the SONiC Routing Working Group. The community has agreed on the following solution: migrating the communication channel between FRR and SONiC from Netlink to Protobuf, with FRR delivering all messages to SONiC using Protobuf. However, since the migration effort is substantial (as it requires recoding all message types supported so far from Netlink to Protobuf), it has been decided to adopt a transitional approach.
+This involves introducing a new communication channel between FRR and SONiC specifically for exchanging information encoded in Protobuf format. In the short term, this new channel operates alongside the existing Netlink-based channel, running in parallel. New features are expected to utilize the Protobuf channel, while legacy functionalities will gradually transition from Netlink to Protobuf. Eventually, all messages will be transmitted through Protobuf.
+
```
Comment (kperumalbfn) [inline doc/fpmsyncd/frr_sonic_protobuf_channel.md:61]: What is the trigger to choose NL or Protobuf format?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+In order to support transmitting information encoded in Protobuf format, we introduce a new module in FRR. The new module is called `dplane_fpm_pb` and is responsible for encoding and sending information to SONiC using Protobuf as the message format. The new FPM module (`dplane_fpm_pb`) is designed to coexist with the existing one (`dplane_fpm_nl`). The two modules can be run simultaneously.
+
+To enable the FPM Protobuf module in FRR, two steps are required:
+
+* FRR must be compiled with the support for the FPM Protobuf module.
+* The FRR zebra daemon must load the FPM Protobuf module during startup.
+
+To implement these modifications, a PR has been submitted to the FRR repository and is currently under review: https://github.com/FRRouting/frr/pull/14173.
+
```
Comment (kperumalbfn) [inline doc/fpmsyncd/frr_sonic_protobuf_channel.md:109]: These changes are not yet merged to FRR. Could you provide the targeted FRR release with protobuf change?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  1. Generate an FPM message containing an SRv6 VPN route encoded using Protobuf.
+  2. Call `FpmLink::processFpmMessage()` with the generated FPM message.
+  3. Verify the delivery of the FPM message to RouteSync by checking if `RouteSync.onMsgRaw()` is invoked.
+  4. Ensure that the route is added to the ROUTE_TABLE of APPL_DB.
+
+* **Test case 2:** Receiving an SRv6 SID encoded using the Protobuf format
+  1. Generate an FPM message containing an SRv6 SID encoded using Protobuf.
+  2. Call `FpmLink::processFpmMessage()` with the generated FPM message.
+  3. Verify the delivery of the FPM message to RouteSync by checking if `RouteSync.onMsgRaw()` is invoked.
+  4. Ensure that the route is added to the SRV6_MY_SID_TABLE of APPL_DB.
```
Comment (kperumalbfn) [inline doc/fpmsyncd/frr_sonic_protobuf_channel.md:232]: Any out of sequence issue when some information of MY_SID_TABLE uses NL and other information like SRV6 SID uses Protobuf?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 4]
```diff
@@ -109,40 +156,29 @@ To implement these modifications, a PR has been submitted to the FRR repository
...
-
-We perform the following changes:
-
-* We extend SONiC's `fpmsyncd` to accept connections from multiple FPM clients. This is required to support receiving information from both FPM Netlink (`dplane_fpm_nl`) and FPM Protobuf (`dplane_fpm_pb`) modules at the same time.
-* We extend SONiC's `fpmsyncd` to parse messages encoded in protobuf.
-* We extend SONiC's `RouteSync` by adding an handler to parse information encoded in Protobuf format, such as SRv6 routes and SIDs.
+We make some changes in some of the SONiC components, namely `FpmLink` and `RouteSync`, which are part of the `fpmsyncd` process. The reason behind these changes is to enable `fpmsyncd` to process Protobuf messages received from FRR and write the necessary information (such as routes and SRv6 SIDs) to the SONiC APPL DB.
```
Comment (arlakshm) [inline doc/fpmsyncd/frr_sonic_protobuf_channel.md:159]: Today, in `fpmsyncd` registers Message handler for `RTM_NEWROUTE` and `RTM_DELROUTE for route messages. Will there be similar mechanism to register message handler for protobuf message types?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 5]
```diff
@@ -77,29 +114,39 @@ This feature requires:
...
 
-In order to support transmitting information encoded in Protobuf format, we introduce a new module in FRR. The new module is called `dplane_fpm_pb` and is responsible for encoding and sending information to SONiC using Protobuf as the message format. The new FPM module (`dplane_fpm_pb`) is designed to coexist with the existing one (`dplane_fpm_nl`). The two modules can be run simultaneously.
+* The FPM Protobuf module has higher priority than the FPM Netlink module.
 
+When FRR needs to transmit information, such as a route, to SONiC, it passes the data to FPM. Subsequently, FPM scans through the list of FPM encoders and selects the first encoder capable of encoding the requested data. Given that the Protobuf encoder holds a higher priority than the Netlink encoder, this means that FPM attempts to encode the information using the Protobuf encoder. If Protobuf is unable to encode the requested data, FPM falls back to the Netlink encoder. Following this, the chosen encoder encodes the information, and finally, FPM sends the resulting message to SONiC.
```
Comment (arlakshm) [inline doc/fpmsyncd/frr_sonic_protobuf_channel.md:147]: does this mean when FPM protobuf is enabled then FRR might not use the netlink encoder for any of the information that needs to syncd to SONiC

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+The lack of flexibility of Netlink in supporting various use cases in SONiC was previously discussed in the SONiC Routing Working Group. The community has agreed on the following solution: migrating the communication channel between FRR and SONiC from Netlink to Protobuf, with FRR delivering all messages to SONiC using Protobuf. However, since the migration effort is substantial (as it requires recoding all message types supported so far from Netlink to Protobuf), it has been decided to adopt a transitional approach.
+This involves introducing a new communication channel between FRR and SONiC specifically for exchanging information encoded in Protobuf format. In the short term, this new channel operates alongside the existing Netlink-based channel, running in parallel. New features are expected to utilize the Protobuf channel, while legacy functionalities will gradually transition from Netlink to Protobuf. Eventually, all messages will be transmitted through Protobuf.
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply doc/fpmsyncd/frr_sonic_protobuf_channel.md]: Hi @arlakshm, many thanks for the review.

As per the last SONIC Routing WG call, we have updated this HLD.
Protobuf is no longer used. All messages sent from FRR to SONiC are encoded in Netlink format.

Could you please review the updated HLD?

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md (hunk 4 above)
Comment (cscarpitta) [reply doc/fpmsyncd/frr_sonic_protobuf_channel.md]: Ditto. In the updated HLD, we are no longer using Protobuf.

File: doc/fpmsyncd/frr_sonic_protobuf_channel.md (hunk 5 above)
Comment (cscarpitta) [reply doc/fpmsyncd/frr_sonic_protobuf_channel.md]: Ditto. In the updated HLD, we are no longer using Protobuf.

**Comment by cscarpitta** [comment]
> @cscarpitta , can you please check the following?
> 
> 1. Add a section in HLD for Warmboot implications.
> 2. Any modifications to existing APP_DB schema with this proposal?
> 3. Any sequencing issues causing dependency since now some are coming via netlink and some via protobuf?

@prsunny

Many thanks for the review.

1. There is no impact on warmboot.
2. No APP_DB schema modifications.
3. We reviewed this proposal in the Routing WG meeting, and I updated the proposal based on the feedback of the community. Now Protobuf is no longer used. All messages are now sent via Netlink.

**Review by zice312963205 — APPROVED** [review]

**Comment by eddieruan-alibaba** [comment]
I could help to merge if no one has any other comments.

**Pull request merged**